sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java

```diff
@@ -558,7 +574,23 @@ long flush(
               appendFailures.inc();
               return RetryType.RETRY_ALL_OPERATIONS;
             },
-            c -> recordsAppended.inc(c.protoRows.getSerializedRowsCount()),
+            c -> {
+              recordsAppended.inc(c.protoRows.getSerializedRowsCount());
+              if (successfulRowsReceiver != null) {
+                for (ByteString rowBytes : c.protoRows.getSerializedRowsList()) {
+                  try {
+                    TableRow row =
+                        TableRowToStorageApiProto.tableRowFromMessage(
+                            DynamicMessage.parseFrom(
+                                Preconditions.checkStateNotNull(appendClientInfo).getDescriptor(),
+                                rowBytes));
+                    successfulRowsReceiver.output(row);
+                  } catch (InvalidProtocolBufferException e) {
+                    LOG.warn("Failure parsing TableRow: " + e);
```

Comment:
Stack trace is lost; would be good to have it. Would be useful to add a custom metric to track these errors, otherwise there is no way to know if these error occur. This failure can also affect the pipeline correctness. If all records fail to convert - none of the windows will show up. A couple of options to make it more robust - carry customer defined "record key" through the whole processing and output that instead of the full collection or output a TableRow with a "failed-to-convert" column/value and the same window as the original.

Reply:
In theory a failure here should be impossible: these bytes are the result of serializing the protocol buffer in the previous stage. The only way I could see this failing is if data on the wire was corrupted, in which case there's not much we can do.

Reply:
Can it happen after the pipeline update?